tools/integration-tests/tester/tests/synchronization_test.go

```diff
@@ -0,0 +1,84 @@
+package tests
+
+import (
+	"testing"
+	"time"
+
+	"github.com/iotaledger/goshimmer/tools/integration-tests/tester/framework"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestNodeSynchronization checks whether messages are synchronized by a peer that joined the network later
+// and initially missed some messages.
+func TestNodeSynchronization(t *testing.T) {
+	n := f.CreateNetwork("TestNodeSynchronization", 4, 2)
+	defer n.Shutdown()
+
+	numMessages := 100
+	ids := make([]string, numMessages)
```

Comment:
Consider using a map as a set so you don't have to loop over to find whether an id exists in `checkForMessageIds`

Reply:
The problem here is that we can't trust the output of `resp.Messages`. Even if `len(resp.Messages)==len(ids)` it doesn't mean it's all the right `ids`. There could be duplicates or even all the same messages. When checking with a map this would still succeed. Therefore, we need to explicitly check for every `id` within `resp.Messages`.

Reply:
What I was referring to is that in `checkForMessageIds` you have a for loop with an inner for loop to check against resp.Messages whether the given ID was one of the ones from the issued messages. If you store the IDs of the messages you've issued into a set, you can simply iterate over resp.Message and check the set whether the given ID exists.

This:
```
	outer:
		for _, id := range ids {
			for _, msg := range resp.Messages {
				// if message found skip to next
				if msg.Id == id {
					continue outer
				}
			}

			t.Errorf("MessageId=%s, issuer=%s not found in peer %s.", id, idsMap[id].issuerPublicKey, peer.String())
		}
```

would become this:
```
for _, msg := range resp.Messages {
           if , has := ids[msg.Id]; !has {
               t.Errorf("MessageId=%s, issuer=%s not found in peer %s.", id, idsMap[id].issuerPublicKey, peer.String())
           }
}
```

Reply:
Yes but iterating over `resp.Message` has exactly the problems I stated above. We can't trust this as this could be x times the same message and then the second version would still succeed. But it should not.

Reply:
When you're adding the id to the set after issuing a message, you simply check whether the id is already in the set and if so, you error out early since a duplicated id was returned from a newly issued message.

Reply:
Changed it to use a map.